cloudify/plugins/workflows.py

```diff
@@ -278,6 +285,139 @@ def validate_inclusions_and_exclusions(include_instances, exclude_instances,
+        rollback_action = modification.rollback
+        if rollback_behavior == ROLLBACK_NONE:
+            raise
+        elif rollback_behavior == ROLLBACK_FAILED:
+            # Only rollback the failed instances not everything
+            rollback_modification = {
+                'modification_action': modification_action,
+                'rollback_instances': [
+                    instance.id for instance in scaled_instances
+                ],
+            }
+            modification.rollback_modification = rollback_modification
+            rollback_action = modification.partial_rollback
+
+        ctx.logger.warn('Rolling back deployment modification. '
+                        '[modification_id={0}]'.format(modification.id))
+        try:
+            rollback_action()
+        except Exception:
+            ctx.logger.warn('Deployment modification rollback failed. The '
+                            'deployment model is most likely in some corrupted'
+                            ' state.'
+                            '[modification_id={0}]'.format(modification.id))
+            raise
+
+    if delta > 0:
+        # get_failed_instances_from_scale_workflow(ctx, )
+        # At this point, the tasks still exist on task graph are the
+        # failed one, so no further actions required
+        ctx.logger.error('Scale out failed, scaling back in.')
+        try:
+            for task in graph.tasks_iter():
+                graph.remove_task(task)
+            lifecycle.uninstall_node_instances(
+                graph=graph,
+                node_instances=scaled_instances,
+                ignore_failure=ignore_failure,
+                related_nodes=related_instances)
+        except Exception:
+            _rollback_modification()
+            raise
+
+    # do the actual rollback
+    _rollback_modification()
+
+
+def get_failed_scaled_instances(ctx,
+                                ready_state,
+                                scaled_instances,
+                                related_instances):
+    """
+    This method will helps to get the failed scaled node instances when do
+    scale out/in
+    :param ctx: cloudify context
+    :param ready_state: A State used in order to tell if the node instances
+    failed or not. In Scale out if the node instance failed then it will not be
+    in "started" state. On the other hand, if the scale in failed in one of
+    the scaled instances then it will not be in "deleted" state. The two
+    possible values for ready_state is "started or deleted"
+    :param scaled_instances: Represent a list of all scaled in/out instances
+    :param related_instances: Represent a list of related instances
+    generated from scale in/out operation
+    :return: Dict contains both failed instances and related node instances
+    """
+    result = {}
+    failed_instances = set()
+    intact_instances = set()
+    for instance in ctx.node_instances:
+        if instance.state != ready_state:
```

Comment:
if you reverse this to `if instance.state == ready_state: continue` then the whole block below will be able to be indented less